gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -423,12 +475,28 @@ private void commitDataset(JobState.DatasetState datasetState) throws IOExceptio
 
     // Set the dataset state to COMMITTED upon successful commit
     datasetState.setState(JobState.RunningState.COMMITTED);
+    return commitSequence;
   }
 
   /**
    * Persist dataset state of a given dataset identified by the dataset URN.
    */
   private void persistDatasetState(String datasetUrn, JobState.DatasetState datasetState) throws IOException {
+    finalizeDatasetState(datasetState, datasetUrn);
+    LOG.info("Persisting dataset state for dataset " + datasetUrn);
+    this.datasetStateStore.persistDatasetState(datasetUrn, datasetState);
+  }
+
+  private Optional<CommitStep> buildDatasetStateCommitStep(String datasetUrn, JobState.DatasetState datasetState)
+      throws IOException {
+
+    finalizeDatasetState(datasetState, datasetUrn);
```

Comment:
Seems you can move `finalizeDatasetState(datasetState, datasetUrn)` out from both `persistDatasetState` and this method into `commit`.